File: leetcode/rangeSumBST.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int rangeSumBST(TreeNode* root, int low, int high) {
        int sum = 0;

        if (root->val >= low && root->val <= high) sum += root->val;

        if (root->left != nullptr) sum += rangeSumBST(root->left, low, high);
        if (root->right != nullptr) sum += rangeSumBST(root->right, low, high);

        return sum;
    }
};

int main() {

    Solution solver;

    TreeNode* a = new TreeNode(10);
    TreeNode* al = new TreeNode(5);
    TreeNode* ar = new TreeNode(15);
    TreeNode* all = new TreeNode(3);
    TreeNode* alr = new TreeNode(7);
    TreeNode* arr = new TreeNode(18);

    a->left = al;
    a->right = ar;

    al->left = all;
    al->right = alr;

    ar->right = arr;

    cout << solver.rangeSumBST(a, 7, 15) << endl;


    return 0;
}

File: leetcode/rangeSumBST.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# https://leetcode.com/problems/range-sum-of-bst/
class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        if not root: return 0
        result = root.val if low <= root.val <= high else 0
        result += self.rangeSumBST(root.left, low, high) if low <= root.val else 0
        result += self.rangeSumBST(root.right, low, high) if root.val <= high else 0
            
        return result
        
    
if __name__ == "__main__":
    # [10,5,15,3,7,null,18]
    l3 = TreeNode(3)
    l7 = TreeNode(7)
    l18 = TreeNode(18)
    n5 = TreeNode(5, l3, l7)
    n15 = TreeNode(15, None, l18)
    r = TreeNode(10, n5, n15)

    sol = Solution()
    print(sol.rangeSumBST(r, 7, 15))
        
        